src/librustc/hir/map/mod.rs

```diff
@@ -497,15 +497,15 @@ impl<'hir> Map<'hir> {
     }
 
     pub fn trait_impls(&self, trait_did: DefId) -> &'hir [NodeId] {
-        self.dep_graph.read(DepNode::TraitImpls(trait_did));
+        self.dep_graph.read(DepNode::AllLocalTraitImpls);
```

Comment:
Mostly for ease of implementation and symmetry with the situation in metadata. `DepNode::TraitImpls` has been re-purposed to represent local *and* remote impls and, with red-green, will take care of stopping invalidation short.

Without red-green though you are right, it's losing precision. Want me to make it per-item?